source/extensions/common/wasm/wasm.h

```diff
@@ -0,0 +1,414 @@
+  Wasm(absl::string_view runtime, absl::string_view vm_id, absl::string_view vm_configuration,
+       Stats::ScopeSharedPtr scope, Upstream::ClusterManager& cluster_manager,
+       Event::Dispatcher& dispatcher);
+  Wasm(const Wasm& other, Event::Dispatcher& dispatcher);
+  ~Wasm();
+
+  bool initialize(const std::string& code, bool allow_precompiled = false);
+  void startVm(Context* root_context);
+  bool configure(Context* root_context, PluginSharedPtr plugin, absl::string_view configuration);
+  Context* start(PluginSharedPtr plugin); // returns the root Context.
+
+  absl::string_view vm_id() const { return vm_id_; }
+  absl::string_view vm_id_with_hash() const { return vm_id_with_hash_; }
+  WasmVm* wasm_vm() const { return wasm_vm_.get(); }
+  Context* vm_context() const { return vm_context_.get(); }
+  Stats::StatNameSetSharedPtr stat_name_set() const { return stat_name_set_; }
+  Context* getRootContext(absl::string_view root_id) { return root_contexts_[root_id].get(); }
+  Context* getContext(uint32_t id) {
+    auto it = contexts_.find(id);
+    if (it != contexts_.end())
+      return it->second;
+    return nullptr;
+  }
+  uint32_t allocContextId();
+
+  Upstream::ClusterManager& clusterManager() const { return cluster_manager_; }
+  const std::string& code() const { return code_; }
+  const std::string& vm_configuration() const { return vm_configuration_; }
+  bool allow_precompiled() const { return allow_precompiled_; }
+  void setInitialConfiguration(const std::string& vm_configuration) {
+    vm_configuration_ = vm_configuration;
+  }
+  Event::Dispatcher& dispatcher() { return dispatcher_; }
+
+  void queueReady(uint32_t root_context_id, uint32_t token);
+
+  void shutdown();
+  WasmResult done(Context* root_context);
+
+  //
+  // AccessLog::Instance
+  //
+  void log(absl::string_view root_id, const Http::HeaderMap* request_headers,
+           const Http::HeaderMap* response_headers, const Http::HeaderMap* response_trailers,
+           const StreamInfo::StreamInfo& stream_info);
+
+  // Support functions.
+  void* allocMemory(uint64_t size, uint64_t* address);
+  // Allocate a null-terminated string in the VM and return the pointer to use as a call arguments.
+  uint64_t copyString(absl::string_view s);
+  uint64_t copyBuffer(const Buffer::Instance& buffer);
+  // Copy the data in 's' into the VM along with the pointer-size pair. Returns true on success.
+  bool copyToPointerSize(absl::string_view s, uint64_t ptr_ptr, uint64_t size_ptr);
+  bool copyToPointerSize(const Buffer::Instance& buffer, uint64_t start, uint64_t length,
+                         uint64_t ptr_ptr, uint64_t size_ptr);
+  template <typename T> bool setDatatype(uint64_t ptr, const T& t);
+
+  // For testing.
+  void setContext(Context* context) { contexts_[context->id()] = context; }
+  void startForTesting(std::unique_ptr<Context> root_context, PluginSharedPtr plugin);
+
+  bool getEmscriptenVersion(uint32_t* emscripten_metadata_major_version,
```

Comment:
This can be used to support multiple emscripten versions.  Currently we are not doing that because we have not upstreamed and so we don't have lots of existing WASM plugins.  Once we upstream and have production users we will want to support previous versions of emscripten subject to a policy which we should develop.